rpc: Add getblockstats

alamshafil opened the pull request:
This PR adds the `getblockstats` RPC which is useful for getting info about a block. 

This RPC is found in `1.21-dev` and I tried to match the output of `1.21-dev` which includes SegWit info and uses virtual bytes (sat/vbyte).

I have tried to use the `getblockstats.json` test data from 1.21-dev but the test fail due to `high-hash` so I have assumed that 1.21-dev test data are SHA-256 Bitcoin hashes (I need to verify this). I have provided test data generated from the `getblockstats.py` qa test.

Adding this RPC was helpful in porting https://mempool.space/ to Dogecoin.

edtubbs commented:
Built for x86_64-linux and `rpc-tests` are passing on Ubuntu.
> I have tried to use the `getblockstats.json` test data from 1.21-dev but the test fail due to `high-hash` so I have assumed that 1.21-dev test data are SHA-256 Bitcoin hashes (I need to verify this). I have provided test data generated from the `getblockstats.py` qa test.

I'll see if I can validate the test data as well and approve once I'm done reviewing the changes.  Looks good!
```
ed@ed-VirtualBox:~/dogecoin/dogecoin$ qa/pull-tester/rpc-tests.py
............................................................
wallet-hd.py:
Pass: True, Duration: 30 s

..........
auxpow.py:
Pass: True, Duration: 5 s

....
p2p-fullblocktest.py:
Pass: True, Duration: 38 s

................
fundrawtransaction.py:
Pass: True, Duration: 47 s

...
getauxblock.py:
Pass: True, Duration: 13 s

.....
wallet-accounts.py:
Pass: True, Duration: 5 s

........
wallet-dump.py:
Pass: True, Duration: 7 s

....................................
listtransactions.py:
Pass: True, Duration: 23 s

............................................
p2p-policy.py:
Pass: True, Duration: 41 s

......
p2p-acceptblock.py:
Pass: True, Duration: 26 s

.........................
wallet.py:
Pass: True, Duration: 75 s

......
zapwallettxes.py:
Pass: True, Duration: 16 s

...............
importmulti.py:
Pass: True, Duration: 11 s


sendheaders.py:
Pass: True, Duration: 28 s

......
walletbackup.py:
Pass: True, Duration: 129 s

.........
mempool_limit.py:
Pass: True, Duration: 17 s

...
merkle_blocks.py:
Pass: True, Duration: 11 s

...........
receivedby.py:
Pass: True, Duration: 16 s

.......
abandonconflict.py:
Pass: True, Duration: 17 s

....
reindex.py:
Pass: True, Duration: 12 s

.......
p2p_invalid_locator.py:
Pass: True, Duration: 4 s

.
p2p-addr.py:
Pass: True, Duration: 11 s

....
mempool_resurrect_test.py:
Pass: True, Duration: 3 s

....
rawtransactions.py:
Pass: True, Duration: 24 s

.............
txn_doublespend.py --mineblock:
Pass: True, Duration: 12 s

..
p2p-tx-download.py:
Pass: True, Duration: 21 s

.
txn_clone.py:
Pass: True, Duration: 12 s

...
mempool_spendcoinbase.py:
Pass: True, Duration: 3 s

...........
mempool_reorg.py:
Pass: True, Duration: 8 s

....
httpbasics.py:
Pass: True, Duration: 8 s

..
multi_rpc.py:
Pass: True, Duration: 4 s

.....
signrawtransactions.py:
Pass: True, Duration: 3 s

.....
rest.py:
Pass: True, Duration: 20 s

....
decodescript.py:
Pass: True, Duration: 2 s

..
proxy_test.py:
Pass: True, Duration: 11 s

.
getchaintips.py:
Pass: True, Duration: 32 s

...
nodehandling.py:
Pass: True, Duration: 10 s


disablewallet.py:
Pass: True, Duration: 3 s

.
blockchain.py:
Pass: True, Duration: 6 s

..
p2p-mempool.py:
Pass: True, Duration: 3 s

....
keypool.py:
Pass: True, Duration: 7 s

.
invalidblockrequest.py:
Pass: True, Duration: 5 s

.........
importprunedfunds.py:
Pass: True, Duration: 6 s

.
preciousblock.py:
Pass: True, Duration: 9 s

...
signmessages.py:
Pass: True, Duration: 3 s

...
prioritise_transaction.py:
Pass: True, Duration: 16 s


createauxblock.py:
Pass: True, Duration: 10 s

...................................
feelimit.py:
Pass: True, Duration: 18 s

.
paytxfee.py:
Pass: True, Duration: 19 s


import-rescan.py:
Pass: True, Duration: 24 s

.....
rpcnamedargs.py:
Pass: True, Duration: 3 s

............
dustlimits.py:
Pass: True, Duration: 31 s

................
p2p-leaktests.py:
Pass: True, Duration: 8 s

......
bumpfee.py:
Pass: True, Duration: 22 s

........
replace-by-fee.py:
Pass: True, Duration: 8 s

........
rescan.py:
Pass: True, Duration: 9 s

.......
listsinceblock.py:
Pass: True, Duration: 31 s

......
wallet_create_tx.py:
Pass: True, Duration: 12 s


getblockstats.py:
Pass: True, Duration: 4 s

......
addnode.py:
Pass: True, Duration: 4 s

...
liststucktransactions.py:
Pass: True, Duration: 13 s

......................
zmq_test.py:
Pass: True, Duration: 17 s

...
setmaxconnections.py:
Pass: True, Duration: 59 s

TEST                           | PASSED | DURATION

wallet-hd.py                   | True   | 30 s
auxpow.py                      | True   | 5 s
p2p-fullblocktest.py           | True   | 38 s
fundrawtransaction.py          | True   | 47 s
getauxblock.py                 | True   | 13 s
wallet-accounts.py             | True   | 5 s
wallet-dump.py                 | True   | 7 s
listtransactions.py            | True   | 23 s
p2p-policy.py                  | True   | 41 s
p2p-acceptblock.py             | True   | 26 s
wallet.py                      | True   | 75 s
zapwallettxes.py               | True   | 16 s
importmulti.py                 | True   | 11 s
sendheaders.py                 | True   | 28 s
walletbackup.py                | True   | 129 s
mempool_limit.py               | True   | 17 s
merkle_blocks.py               | True   | 11 s
receivedby.py                  | True   | 16 s
abandonconflict.py             | True   | 17 s
reindex.py                     | True   | 12 s
p2p_invalid_locator.py         | True   | 4 s
p2p-addr.py                    | True   | 11 s
mempool_resurrect_test.py      | True   | 3 s
rawtransactions.py             | True   | 24 s
txn_doublespend.py --mineblock | True   | 12 s
p2p-tx-download.py             | True   | 21 s
txn_clone.py                   | True   | 12 s
mempool_spendcoinbase.py       | True   | 3 s
mempool_reorg.py               | True   | 8 s
httpbasics.py                  | True   | 8 s
multi_rpc.py                   | True   | 4 s
signrawtransactions.py         | True   | 3 s
rest.py                        | True   | 20 s
decodescript.py                | True   | 2 s
proxy_test.py                  | True   | 11 s
getchaintips.py                | True   | 32 s
nodehandling.py                | True   | 10 s
disablewallet.py               | True   | 3 s
blockchain.py                  | True   | 6 s
p2p-mempool.py                 | True   | 3 s
keypool.py                     | True   | 7 s
invalidblockrequest.py         | True   | 5 s
importprunedfunds.py           | True   | 6 s
preciousblock.py               | True   | 9 s
signmessages.py                | True   | 3 s
prioritise_transaction.py      | True   | 16 s
createauxblock.py              | True   | 10 s
feelimit.py                    | True   | 18 s
paytxfee.py                    | True   | 19 s
import-rescan.py               | True   | 24 s
rpcnamedargs.py                | True   | 3 s
dustlimits.py                  | True   | 31 s
p2p-leaktests.py               | True   | 8 s
bumpfee.py                     | True   | 22 s
replace-by-fee.py              | True   | 8 s
rescan.py                      | True   | 9 s
listsinceblock.py              | True   | 31 s
wallet_create_tx.py            | True   | 12 s
getblockstats.py               | True   | 4 s
addnode.py                     | True   | 4 s
liststucktransactions.py       | True   | 13 s
zmq_test.py                    | True   | 17 s
setmaxconnections.py           | True   | 59 s

ALL                            | True   | 1105 s (accumulated)

Runtime: 288 s
ed@ed-VirtualBox:~/dogecoin/dogecoin$
```

alamshafil commented on src/rpc/blockchain.cpp:
```diff
@@ -1481,10 +1513,324 @@ UniValue reconsiderblock(const JSONRPCRequest& request)
+
+    UniValue ret_all(UniValue::VOBJ);
+    ret_all.pushKV("avgfee", (block.vtx.size() > 1) ? totalfee / (block.vtx.size() - 1) : 0);
+    ret_all.pushKV("avgfeerate", total_weight ? (totalfee * WITNESS_SCALE_FACTOR) / total_weight : 0); // Unit: sat/vbyte
```
Do front-ends use DOGE/kb instead of koinu/byte?

alamshafil commented:
I have removed segwit info and switched "sat/vByte" to "koinu/byte".

Regrading the subsidy calculation, I was not sure about the `sum()` part so I assumed it was to add all outputs for the first TX. Please correct if this is wrong or has any issue regrading the types.

```cpp

    int64_t subsidy = 0;
    for (const auto& tx: block.vtx[0]->vout) {
        subsidy += tx.nValue;
    }
    subsidy = std::max(subsidy - totalfee, (int64_t)0);
```

chromatic commented:
I'll review this against #3248 to see if it meets my needs. If so, or if this can be extended to meet my needs, I'm happy to defer my PR in favor of this one.

alamshafil commented:
Rebased on top of 1.14.7-dev and updated to use `GetCheckedBlock()`

patricklodder reviewed (commented):
> Note: I'm getting an error when testing with `dogecoin-cli` and using hash instead of height:
> 
> ```
> $ src/dogecoin-cli -testnet getblockstats $(src/dogecoin-cli -testnet getbestblockhash)
> error: Error parsing JSON:028e2441614c54a9557df1ddebf4809a86df56214c8858e158023da177955d36
> ```

I'm still getting this error. It is caused by the first parameter being either a JSON Number or String. This is hellish to implement for any typed language, including `dogecoin-cli`. Currently I have to do:

```console
$ src/dogecoin-cli -testnet getblockstats "\"$(src/dogecoin-cli -testnet getbestblockhash)\""
```

to make the cli work, which is truly ugly. Therefore I'm considering this a bad API spec.

Can we choose either of (in order of my preference but we can discuss this):
1. Support only block hash, like with `getblock`, or
2. Make a separate API call `getblockstatsbyheight` and call the current logic from each call handler, or
3. Separate the parameters (at the cost of forcing named parameters or more ugliness)

patricklodder commented on src/rpc/blockchain.cpp:
```diff
@@ -1498,10 +1531,313 @@ UniValue reconsiderblock(const JSONRPCRequest& request)
+
+static UniValue getblockstats(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 4) {
```
Only argument 0 and 1 are used, so the maximum number of arguments given should be 2.

patricklodder commented on src/rpc/blockchain.cpp:
```diff
@@ -1498,10 +1531,313 @@ UniValue reconsiderblock(const JSONRPCRequest& request)
+    LOCK(cs_main);
+
+    CBlockIndex* pindex;
+    if (request.params[0].isNum()) {
+        const int height = request.params[0].get_int();
+        const int current_tip = chainActive.Height();
+        if (height < 0) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Target block height %d is negative", height));
+        }
+        if (height > current_tip) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Target block height %d after current tip %d", height, current_tip));
+        }
+
+        pindex = chainActive[height];
+    } else {
```
Per earlier comment about type-safety, recommend this to be removed and keep this method hash-only.

See https://github.com/patricklodder/dogecoin/commit/59dcc34e9c506ea1cc2f846aeb3dc35da14f636e for a proposal to implement that.

alamshafil commented:
I have updated this pull request with the proposal from https://github.com/patricklodder/dogecoin/commit/59dcc34e9c506ea1cc2f846aeb3dc35da14f636e.

patricklodder reviewed (commented):
I've retested this. Could use a squash. I'm fine if you squash my nits into your commit.

Needs another maintainer to approve because I co-authored this now.

alamshafil commented:
PR has been squashed.

patricklodder reviewed (approved):
ACK - tested on Ubuntu 22.04 x86_64, cross-compiled to linux x86_64 and aarch64, and macOS Ventura arm64 - I've been running this on both testnet and mainnet for almost 3 months now.

Needs 1 more approval because I contributed some code to this via #3393

chromatic commented on src/rpc/blockchain.cpp:
```diff
@@ -729,6 +734,43 @@ UniValue getblockheader(const JSONRPCRequest& request)
+static CBlock GetBlockChecked(const CBlockIndex* pblockindex)
+{
+    CBlock block;
+    if (fHavePruned && !(pblockindex->nStatus & BLOCK_HAVE_DATA) && pblockindex->nTx > 0) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
+    }
+
+    if (!ReadBlockFromDisk(block, pblockindex, Params().GetConsensus(pblockindex->nHeight))) {
+        // Block not found on disk. This could be because we have the block
+        // header in our index but don't have the block (for example if a
+        // non-whitelisted node sends us an unrequested long chain of valid
+        // blocks, we add the headers to our index, but don't accept the
+        // block).
+        throw JSONRPCError(RPC_MISC_ERROR, "Block not found on disk");
```
It'd be nice to test for these conditions, though it shouldn't block this PR (and I imagine they might be somewhat tricky to test).

chromatic commented on src/rpc/blockchain.cpp:
```diff
@@ -1498,10 +1531,300 @@ UniValue reconsiderblock(const JSONRPCRequest& request)
+
+    LOCK(cs_main);
+
+    CBlockIndex* pindex;
+    const uint256 hash = ParseHashV(request.params[0], "hash");
+
+    if (mapBlockIndex.count(hash) == 0)
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
+    pindex = mapBlockIndex[hash];
+
+    if (!chainActive.Contains(pindex)) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Block is not in chain %s", Params().NetworkIDString()));
+    }
```
Same note about tests here, though again not a blocking condition for this PR.

patricklodder replied:
The on line 1651 is tested here: https://github.com/dogecoin/dogecoin/pull/3297/files#diff-1ce3752d494e0fc0cbe3d60b2c29ba8bc29cb9ccd18b74ba7c0186cf8dbf8ebeR166

The bottom one can be tested from a chaintip fork: `dogecoin-cli getchaintips` and take any forked tip with "valid-headers" / "valid-forks".

e.g.:

```console
$ dogecoin-cli -testnet getblockstats 49841b077786daf0a97dd848246b74635be01b1d99447c79d1f668c31e646d70
error code: -8
error message:
Block is not in chain test
```

chromatic reviewed (approved):
ACK - tests pass on 64-bit x86 Linux and code is fine. I left a couple of comments for potential future enhancements.

Do we want to mark this command as experimental or subject to change? I'd like to see how it's used for a while to see if there are any usability issues we might want to address, but I don't feel strongly about that (and we can make that happen in a separate PR if we agree).

patricklodder commented:
> Do we want to mark this command as experimental or subject to change?

We have #3212 open to document exactly that - I'll pick compiling this up tomorrow.

slightlyskepticalpotat commented:
`rpc-tests` also passing on my end with 64-bit Ubuntu 22.04.
